### PR TITLE
Don't emit CA1064 in executables

### LINKED ISSUE #7191: Conflict between CA1515 and CA1064 for Exception types
### Analyzers

* [CA1515](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1515): `Consider making public types internal`
* [CA1064](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1064): `Exceptions should be public`

### Analyzer source

**Version**: [SDK 9.0.100-preview.1](https://dotnet.microsoft.com/en-us/download/dotnet/9.0)

### Describe the bug

Setting the visibility of an `Exception` class to `public` causes a CA1515 warning, whereas setting it to `internal` causes a CA1064 warning.

### Steps To Reproduce

**ConsoleApp1.csproj**

```csproj
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net9.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
  </PropertyGroup>
</Project>
```

**Program.cs**
```cs
throw new ExampleException("Test");

public sealed class ExampleException : Exception // or internal
{
  public ExampleException(string message) : base(message) { }
  public ExampleException(string message, Exception innerException) : base(message, innerException) { }
}
```

### Expected behavior
One of the visibility modifiers (`public` or `internal`) should compile without either compiler warning.

### Actual behavior
Both options result in a compiler warning.

### PR DESCRIPTION
**Pull request by CollinAlpert**
Affected analyzers: MakeTypesInternal, ExceptionsShouldBePublicAnalyzer
Affected diagnostic IDs: [CA1515](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1515), [CA1064](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/CA1064)

This PR ensures that exception types are not suggested to be made internal, as this could conflict with [CA1064](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1064).

Fixes #7191

### CODE REVIEW

**Review by buyaa-n — CHANGES_REQUESTED** [review]
Thank you @CollinAlpert for the fix, as [discussed in the issue](https://github.com/dotnet/roslyn-analyzers/issues/7191#issuecomment-2153878116) we don't want to exclude exception types from CA1515, instead it's better to exclude diagnostic for [CA1064 Exceptions should be public:](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1064) in executable. 

- Please revert the `Exclude exception types from CA1515` part from the PR, looks you also did some code refactoring for the analyzer we can accept that part.
- If you are interested fixing the issue by excluding CA1064 diagnostics for executable, that will be great, but not required

**Comment by CollinAlpert** [comment]
Thanks for the hint, I have reverted the change to CA1515 and tweaked CA1064.

**Review by buyaa-n — COMMENTED** [review]
Left a few comments, overall looks great, thank you @CollinAlpert!

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/MakeTypesInternal.cs [hunk 1]
```diff
@@ -42,23 +41,22 @@ public override void Initialize(AnalysisContext context)
                     return;
```
[the remark this answers is not in the document]
Comment (CollinAlpert) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/MakeTypesInternal.cs:41]: `MakeTypesInternal` allows configuring the `OutputKind`s via `.editorconfig`, something `ExceptionsShouldBePublicAnalyzer` does not. This is why the check is a bit more complicated.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/MakeTypesInternal.cs [hunk 2]
```diff
@@ -42,23 +41,22 @@ public override void Initialize(AnalysisContext context)
...
-        protected abstract ImmutableArray<TSyntaxKind> TypeKinds { get; }
-
-        protected abstract TSyntaxKind EnumKind { get; }
-
-        protected abstract ImmutableArray<TSyntaxKind> DelegateKinds { get; }
-
-        protected abstract void AnalyzeTypeDeclaration(SyntaxNodeAnalysisContext context);
-
-        protected abstract void AnalyzeEnumDeclaration(SyntaxNodeAnalysisContext context);
+        /// <see cref="Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.ExceptionsShouldBePublicAnalyzer"/>
```
[the remark this answers is not in the document]
Comment (CollinAlpert) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/MakeTypesInternal.cs]: Good point, I removed it.

**Review by buyaa-n — APPROVED** [review]
Thanks @CollinAlpert looks good

**Pull request merged**